// @tanstack/start-vite-plugin@1.82.5/dist/cjs/compilers.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const babel = require("@babel/core");
const t = require("@babel/types");
const _generate = require("@babel/generator");
const codeFrame = require("@babel/code-frame");
const babelDeadCodeElimination = require("babel-dead-code-elimination");
const ast = require("./ast.cjs");
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const babel__namespace = /* @__PURE__ */ _interopNamespaceDefault(babel);
const t__namespace = /* @__PURE__ */ _interopNamespaceDefault(t);
let generate = _generate["default"];
if (!generate) {
  generate = _generate;
}
function compileEliminateDeadCode(opts) {
  const ast$1 = ast.parseAst(opts);
  if (!ast$1) {
    throw new Error(
      `Failed to compile ast for compileEliminateDeadCode() for the file: ${opts.filename}`
    );
  }
  babelDeadCodeElimination.deadCodeElimination(ast$1);
  return generate(ast$1, {
    sourceMaps: true
  });
}
const debug = process.env.TSR_VITE_DEBUG === "true";
function compileStartOutput(opts) {
  const ast$1 = ast.parseAst(opts);
  if (!ast$1) {
    throw new Error(
      `Failed to compile ast for compileStartOutput() for the file: ${opts.filename}`
    );
  }
  babel__namespace.traverse(ast$1, {
    Program: {
      enter(programPath) {
        const identifiers = {
          createServerFn: {
            name: "createServerFn",
            type: "ImportSpecifier",
            namespaceId: "",
            handleCallExpression: handleCreateServerFnCallExpression,
            paths: []
          },
          createMiddleware: {
            name: "createMiddleware",
            type: "ImportSpecifier",
            namespaceId: "",
            handleCallExpression: handleCreateMiddlewareCallExpression,
            paths: []
          },
          serverOnly: {
            name: "serverOnly",
            type: "ImportSpecifier",
            namespaceId: "",
            handleCallExpression: handleServerOnlyCallExpression,
            paths: []
          }
        };
        const identifierKeys = Object.keys(identifiers);
        programPath.traverse({
          ImportDeclaration: (path) => {
            if (path.node.source.value !== "@tanstack/start") {
              return;
            }
            path.node.specifiers.forEach((specifier) => {
              identifierKeys.forEach((identifierKey) => {
                const identifier = identifiers[identifierKey];
                if (specifier.type === "ImportSpecifier" && specifier.imported.type === "Identifier") {
                  if (specifier.imported.name === identifierKey) {
                    identifier.name = specifier.local.name;
                    identifier.type = "ImportSpecifier";
                  }
                }
                if (specifier.type === "ImportNamespaceSpecifier") {
                  identifier.type = "ImportNamespaceSpecifier";
                  identifier.namespaceId = specifier.local.name;
                  identifier.name = `${identifier.namespaceId}.${identifierKey}`;
                }
              });
            });
          },
          CallExpression: (path) => {
            identifierKeys.forEach((identifierKey) => {
              var _a;
              if (t__namespace.isIdentifier(path.node.callee) && path.node.callee.name === identifiers[identifierKey].name) {
                if (((_a = path.scope.getBinding(identifiers[identifierKey].name)) == null ? void 0 : _a.path.node.type) === "FunctionDeclaration") {
                  return;
                }
                return identifiers[identifierKey].paths.push(path);
              }
              if (t__namespace.isMemberExpression(path.node.callee)) {
                if (t__namespace.isIdentifier(path.node.callee.object) && t__namespace.isIdentifier(path.node.callee.property)) {
                  const callname = [
                    path.node.callee.object.name,
                    path.node.callee.property.name
                  ].join(".");
                  if (callname === identifiers[identifierKey].name) {
                    identifiers[identifierKey].paths.push(path);
                  }
                }
              }
              return;
            });
          }
        });
        identifierKeys.forEach((identifierKey) => {
          identifiers[identifierKey].paths.forEach((path) => {
            identifiers[identifierKey].handleCallExpression(
              path,
              opts
            );
          });
        });
      }
    }
  });
  return generate(ast$1, {
    sourceMaps: true,
    minified: process.env.NODE_ENV === "production"
  });
}
function handleCreateServerFnCallExpression(path, opts) {
  var _a;
  const calledOptions = path.node.arguments[0] ? path.get("arguments.0") : null;
  const shouldValidateClient = !!(calledOptions == null ? void 0 : calledOptions.node.properties.find((prop) => {
    return t__namespace.isObjectProperty(prop) && t__namespace.isIdentifier(prop.key) && prop.key.name === "validateClient" && t__namespace.isBooleanLiteral(prop.value) && prop.value.value === true;
  }));
  const callExpressionPaths = {
    middleware: null,
    validator: null,
    handler: null
  };
  const validMethods = Object.keys(callExpressionPaths);
  const rootCallExpression = getRootCallExpression(path);
  if (debug)
    console.info(
      "Handling createServerFn call expression:",
      rootCallExpression.toString()
    );
  if (!rootCallExpression.parentPath.isVariableDeclarator()) {
    throw new Error("createServerFn must be assigned to a variable!");
  }
  const variableDeclarator = rootCallExpression.parentPath.node;
  const existingVariableName = variableDeclarator.id.name;
  rootCallExpression.traverse({
    MemberExpression(memberExpressionPath) {
      if (t__namespace.isIdentifier(memberExpressionPath.node.property)) {
        const name = memberExpressionPath.node.property.name;
        if (validMethods.includes(name) && memberExpressionPath.parentPath.isCallExpression()) {
          callExpressionPaths[name] = memberExpressionPath.parentPath;
        }
      }
    }
  });
  if (callExpressionPaths.validator) {
    const innerInputExpression = callExpressionPaths.validator.node.arguments[0];
    if (!innerInputExpression) {
      throw new Error(
        "createServerFn().validator() must be called with a validator!"
      );
    }
    if (opts.env === "client" && !shouldValidateClient && t__namespace.isMemberExpression(callExpressionPaths.validator.node.callee)) {
      callExpressionPaths.validator.replaceWith(
        callExpressionPaths.validator.node.callee.object
      );
    }
  }
  const handlerFnPath = (_a = callExpressionPaths.handler) == null ? void 0 : _a.get(
    "arguments.0"
  );
  if (!callExpressionPaths.handler || !handlerFnPath.node) {
    throw codeFrameError(
      opts.code,
      path.node.callee.loc,
      `createServerFn must be called with a "handler" property!`
    );
  }
  const handlerFn = handlerFnPath.node;
  removeUseServerDirective(handlerFnPath);
  handlerFnPath.replaceWith(
    t__namespace.arrowFunctionExpression(
      [t__namespace.identifier("opts")],
      t__namespace.blockStatement(
        // Everything in here is server-only, since the client
        // will strip out anything in the 'use server' directive.
        [
          t__namespace.returnStatement(
            t__namespace.callExpression(
              t__namespace.identifier(`${existingVariableName}.__executeServer`),
              [t__namespace.identifier("opts")]
            )
          )
        ],
        [t__namespace.directive(t__namespace.directiveLiteral("use server"))]
      )
    )
  );
  if (opts.env === "server") {
    callExpressionPaths.handler.node.arguments.push(handlerFn);
  }
}
function removeUseServerDirective(path) {
  path.traverse({
    Directive(path2) {
      if (path2.node.value.value === "use server") {
        path2.remove();
      }
    }
  });
}
function handleCreateMiddlewareCallExpression(path, opts) {
  var _a;
  const rootCallExpression = getRootCallExpression(path);
  if (debug)
    console.info(
      "Handling createMiddleware call expression:",
      rootCallExpression.toString()
    );
  const callExpressionPaths = {
    middleware: null,
    validator: null,
    client: null,
    server: null
  };
  const validMethods = Object.keys(callExpressionPaths);
  rootCallExpression.traverse({
    MemberExpression(memberExpressionPath) {
      if (t__namespace.isIdentifier(memberExpressionPath.node.property)) {
        const name = memberExpressionPath.node.property.name;
        if (validMethods.includes(name) && memberExpressionPath.parentPath.isCallExpression()) {
          callExpressionPaths[name] = memberExpressionPath.parentPath;
        }
      }
    }
  });
  if (callExpressionPaths.validator) {
    const innerInputExpression = callExpressionPaths.validator.node.arguments[0];
    if (!innerInputExpression) {
      throw new Error(
        "createMiddleware().validator() must be called with a validator!"
      );
    }
    if (opts.env === "client") {
      if (t__namespace.isMemberExpression(callExpressionPaths.validator.node.callee)) {
        callExpressionPaths.validator.replaceWith(
          callExpressionPaths.validator.node.callee.object
        );
      }
    }
  }
  const useFnPath = (_a = callExpressionPaths.server) == null ? void 0 : _a.get(
    "arguments.0"
  );
  if (!callExpressionPaths.server || !useFnPath.node) {
    throw new Error('createMiddleware must be called with a "use" property!');
  }
  if (opts.env === "client") {
    if (t__namespace.isMemberExpression(callExpressionPaths.server.node.callee)) {
      callExpressionPaths.server.replaceWith(
        callExpressionPaths.server.node.callee.object
      );
    }
  }
}
function handleServerOnlyCallExpression(path, opts) {
  if (debug)
    console.info("Handling serverOnly call expression:", path.toString());
  if (opts.env === "server") {
    return;
  }
  path.replaceWith(
    t__namespace.arrowFunctionExpression(
      [],
      t__namespace.blockStatement([
        t__namespace.expressionStatement(
          t__namespace.callExpression(t__namespace.identifier("invariant"), [
            t__namespace.booleanLiteral(false),
            t__namespace.stringLiteral(
              "serverOnly() functions can only be called on the server!"
            )
          ])
        )
      ])
    )
  );
}
function getRootCallExpression(path) {
  let rootCallExpression = path;
  while (rootCallExpression.parentPath.isMemberExpression()) {
    const parent = rootCallExpression.parentPath;
    if (parent.parentPath.isCallExpression()) {
      rootCallExpression = parent.parentPath;
    }
  }
  return rootCallExpression;
}
function codeFrameError(code, loc, message) {
  const frame = codeFrame.codeFrameColumns(
    code,
    {
      start: loc.start,
      end: loc.end
    },
    {
      highlightCode: true,
      message
    }
  );
  return new Error(frame);
}
exports.compileEliminateDeadCode = compileEliminateDeadCode;
exports.compileStartOutput = compileStartOutput;
