// @tanstack/start-vite-plugin@1.82.5/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const node_url = require("node:url");
const compilers = require("./compilers.cjs");
const debug = Boolean(process.env.TSR_VITE_DEBUG);
function TanStackStartViteServerFn(opts) {
  let ROOT = process.cwd();
  return {
    name: "vite-plugin-tanstack-start-create-server-fn",
    enforce: "pre",
    configResolved: (config) => {
      ROOT = config.root;
    },
    transform(code, id) {
      const url = node_url.pathToFileURL(id);
      url.searchParams.delete("v");
      id = node_url.fileURLToPath(url).replace(/\\/g, "/");
      const includesToken = /createServerFn|createMiddleware|serverOnly/.test(
        code
      );
      const includesEitherFunc = /(function createServerFn|function createMiddleware|function serverOnly)/.test(
        code
      );
      if (!includesToken || includesEitherFunc) {
        return null;
      }
      if (code.includes("@react-refresh")) {
        throw new Error(
          `We detected that the '@vitejs/plugin-react' was passed before '@tanstack/start-vite-plugin'. Please make sure that '@tanstack/router-vite-plugin' is passed before '@vitejs/plugin-react' and try again: 
e.g.

plugins: [
  TanStackStartVite(), // Place this before viteReact()
  viteReact(),
]
`
        );
      }
      const compiled = compilers.compileStartOutput({
        code,
        root: ROOT,
        filename: id,
        env: opts.env
      });
      if (debug) console.info("");
      if (debug) console.info("Compiled createServerFn Output");
      if (debug) console.info("");
      if (debug) console.info(compiled.code);
      if (debug) console.info("");
      if (debug) console.info("");
      if (debug) console.info("");
      return compiled;
    }
  };
}
function TanStackStartViteDeadCodeElimination(opts) {
  let ROOT = process.cwd();
  return {
    name: "vite-plugin-tanstack-start-dead-code-elimination",
    enforce: "post",
    configResolved: (config) => {
      ROOT = config.root;
    },
    transform(code, id) {
      const url = node_url.pathToFileURL(id);
      url.searchParams.delete("v");
      id = node_url.fileURLToPath(url).replace(/\\/g, "/");
      if (code.includes("createServerFn") || code.includes("createMiddleware")) {
        const compiled = compilers.compileEliminateDeadCode({
          code,
          root: ROOT,
          filename: id,
          env: opts.env
        });
        if (debug) console.info("");
        if (debug) console.info("Output after dead code elimination");
        if (debug) console.info("");
        if (debug) console.info(compiled.code);
        if (debug) console.info("");
        if (debug) console.info("");
        if (debug) console.info("");
        return compiled;
      }
      return null;
    }
  };
}
exports.TanStackStartViteDeadCodeElimination = TanStackStartViteDeadCodeElimination;
exports.TanStackStartViteServerFn = TanStackStartViteServerFn;
