// @clerk/tanstack-start@0.5.3/dist/server/constants.js
import { apiUrlFromPublishableKey } from '@clerk/shared/apiUrlFromPublishableKey';
import { getPublicEnvVariables, getEnvVariable } from '../utils/env.js';

// src/server/constants.ts
var CLERK_JS_VERSION = getPublicEnvVariables().clerkJsVersion || "";
var CLERK_JS_URL = getPublicEnvVariables().clerkJsUrl || "";
var API_VERSION = getEnvVariable("CLERK_API_VERSION") || "v1";
var SECRET_KEY = getEnvVariable("CLERK_SECRET_KEY") || "";
var PUBLISHABLE_KEY = getPublicEnvVariables().publishableKey || "";
var ENCRYPTION_KEY = getEnvVariable("CLERK_ENCRYPTION_KEY") || "";
var API_URL = getEnvVariable("CLERK_API_URL") || apiUrlFromPublishableKey(PUBLISHABLE_KEY);
var DOMAIN = getPublicEnvVariables().domain || "";
var PROXY_URL = getPublicEnvVariables().proxyUrl || "";
var CLERK_JWT_KEY = getEnvVariable("CLERK_JWT_KEY") || "";
var IS_SATELLITE = getPublicEnvVariables().isSatellite || false;
var SIGN_IN_URL = getPublicEnvVariables().signInUrl || "";
var SIGN_UP_URL = getPublicEnvVariables().signUpUrl || "";
var SDK_METADATA = {
  name: "@clerk/tanstack-start",
  version: "0.5.3",
  environment: getEnvVariable("NODE_ENV")
};
var TELEMETRY_DISABLED = getPublicEnvVariables().telemetryDisabled;
var TELEMETRY_DEBUG = getPublicEnvVariables().telemetryDebug;

export { API_URL, API_VERSION, CLERK_JS_URL, CLERK_JS_VERSION, CLERK_JWT_KEY, DOMAIN, ENCRYPTION_KEY, IS_SATELLITE, PROXY_URL, PUBLISHABLE_KEY, SDK_METADATA, SECRET_KEY, SIGN_IN_URL, SIGN_UP_URL, TELEMETRY_DEBUG, TELEMETRY_DISABLED };
